### DISCUSSION TITLE
🌀Teleporter🌀 script

### DISCUSSION THREAD

**Discussion by euu2021** [freeplane/freeplane#1011] [Show and tell]
With the help of @lilive (see [discussion](https://github.com/lilive/Freeplane-Jumper/issues/13)) I created a script that moves the selected nodes to other spots on the map:
![javaw_FTazHmp1jZ](https://user-images.githubusercontent.com/77707706/219656265-b9f95bc3-1adb-46c7-9b21-253dd13c05bf.gif)

In the example above, the user takes the tasks in his Inbox, and sends it to the appropriate projects.

The script uses the Jumper Add-On to find the target node. 

I used the Move action, but it can be personalized to perform other actions like: connect; clone; copy; link. In the discussion, lilive explains how to use other actions.

The script:

```
EDIT:  see latest version of the script in the comments
```

For a future project: it would be nice to have a script that opens a GUI that allows the user to choose the action to be performed.

**Comment by bepolymathe** [comment]
Oh! That's a great idea. Thank you. 

> I used the Move action, but it can be personalized to perform other actions like: connect; clone; copy; link.

I think I will need all these variations ;-)

**Comment by euu2021** [comment]
I noticed that there is one relevant problem in the script, though: if the user selects a whole branch and teleport it, the nodes will be all at the same level (ie, they will all be moved as a child of the target node). 

As a quick fix, I tried to remove each node that already had an ancestor in the list of nodes to be moved, but it doesn't work:

```
nodeList.each{
    def ancestors = [ ]
    ancestors = it.pathToRoot.collect()
    ancestors -= it
    if(nodeList.any() == ancestors.any()){
        nodeList -= it
    }
}
```

I won't be able to find a solution now, so be careful when using the script. Only use it on the parent of a branch.

**Reply by euu2021** [reply]
Ok, this seems to work fine:

```
// @ExecutionModes({on_single_node})

originalNodePosition = node.parent.getChildPosition(node)
originalParent = node.parent

def nodeList = [ ]
nodeList = c.selecteds.collect()

lilive.jumper.Jumper.start()

if( nodeList.contains(c.getSelected()) ) return

nodeList.each{
    def descendants = [ ]
    descendants = it.findAll().collect()
    descendants -= it
    descendants.each{
        if(nodeList.contains(it)){
            nodeList -= it
        }
    }
}

nodeList.each{it.moveTo(c.selected)}

if (originalParent.children[0] == null) {
    c.select(originalParent)
}
    else if (originalParent.children[originalNodePosition + 1] == null) {
    lastChild = originalParent.getChildren().size()
    c.select(originalParent.children[lastChild - 1])
}
    else {
    c.select(originalParent.children[originalNodePosition])
}

```

**Reply by grotesque** [reply]
not working for me

**Reply by euu2021** [reply]
> not working for me

I will take a look, when I find some time.
